### ISSUE TITLE
Opaque definition no longer strictly-positive

### ISSUE THREAD

**Issue by flupe** [agda/agda#6970]
Defining an alias to a strictly-positive functor as opaque prevents it from being seen as strictly-positive. Is this intended behavior?

```agda
module opaquepos where

open import Agda.Builtin.List

opaque

  MyList : Set → Set
  MyList = List

data A : Set where
  node : List A → A

data B : Set where
  node : MyList B → B
```

```
opaquepos.agda:12,1-13,22
B is not strictly positive, because it occurs
in the first argument of MyList
in the type of the constructor node
in the definition of B.
```

**Comment by jespercockx** [comment]
@plt-amy is this intentional? Could you please take a look?

**Comment by plt-amy** [comment]
This also happens if `MyList` is marked `abstract`. I think it's perfectly sensible: Agda doesn't track polarities in types (yet), so as far as the positivity checker is concerned, `MyList` is _Just Some `Set → Set`_. But I suppose I could be talked down from this position, and the fix is simple..

**Comment by andreasabel** [comment]
Maybe it makes sense for `abstract`, which is kind of trying to prevent any information about the abstract thing to leak (but no one uses it for that purpose).
If `opaque` is an explicit mechanism mainly to prevent unwanted unfoldings, it could allow positivity information to get out...

**Comment by plt-amy** [comment]
Like I said, my view could be swayed! I don't think it makes much sense for the positivity checker to see through opaque definitions like this, but since we don't support opaque data types (hence there would be no way to unfold `MyList` in the definition of `B`), it would be a sensible stop-gap.

**Comment by flupe** [comment]
I agree that the positivity checker should _probably_ not know this by default, and positivity information should ideally be expressed intentionally in the type of opaque definitions with polarity annotations. Since they are not available (yet), an alternative could be to repurpose the `POLARITY` pragma for opaque definitions to explicitly decide whether to expose this information to the positivity checker.

```agda
opaque
  MyList : Set → Set
  MyList = List

  {-# POLARITY MyList ++ #-}
```

Could be coupled with a check that enforces that the given polarities do hold, in order to still allow this with `--safe`.

**Comment by gallais** [comment]
Similar issue with the emptyness check not seeing through `abstract` / `opaque` blocks.
I can see arguments going either ways: why should it be possible to rely on this potentially
'accidental' property but also why shouldn't we be able to make it part of the declared signature?

And that is perhaps the answer: can we come up with surface syntax allowing users to declare
their intent? Test case (currently not working):


```agda
data Empty : Set where

opaque

  ⊥ : Set
  ⊥ = Empty


⊥-elim : ⊥ → ∀ {a} {A : Set a} → A
⊥-elim ()
```

**Comment by andreasabel** [comment]
The OP could probably be fixed by declaring `MyList : ++Set -> Set` after merging:
- #6385

**Comment by andreasabel** [comment]
@gallais: Your example (emptiness check) is working-as-intended, as `\bot` is essentially a postulate unless explicitly unfolded.

**Comment by nad** [comment]
> The OP could probably be fixed by declaring `MyList : ++Set -> Set` after merging:
> 
> - #6385

This does not quite work. I opened a new issue for that (#7795).
